### PR TITLE
Refactor Compiler Plugin Service Validation

### PR DESCRIPTION
**Pull request by ThisaruGuruge**
## Purpose

* Refactoring the compiler plugin validations
* Moving common methods to utils class to use in compile-time schema generation
* Fix a bug that allowed passing incorrect parameter permutations
* Add tests for different listener init combinations

Fixes: [#2733](https://github.com/ballerina-platform/ballerina-standard-library/issues/2733)
Fixes: [#2746](https://github.com/ballerina-platform/ballerina-standard-library/issues/2746)

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests
- [ ] ~Updated the spec~

### CODE REVIEW

File: compiler-plugin-tests/src/test/resources/ballerina_sources/invalid_service_15/service.bal [hunk 1]
```diff
@@ -17,12 +17,35 @@
...
 
 service graphql:Service on listener1 {
     resource function get name() returns string {
             return "John";
     }
 }
+
+service graphql:Service on listener2 {
+    resource function get name() returns string {
+            return "John";
```
Comment (DimuthuMadushan) [inline compiler-plugin-tests/src/test/resources/ballerina_sources/invalid_service_15/service.bal:37]:
```suggestion
        return "John";
```

File: compiler-plugin-tests/src/test/resources/ballerina_sources/invalid_service_15/service.bal [hunk 2]
```diff
@@ -17,12 +17,35 @@
...
+    }
+}
+
+service graphql:Service on listener3 {
+    resource function get name() returns string {
+            return "John";
```
Comment (DimuthuMadushan) [inline compiler-plugin-tests/src/test/resources/ballerina_sources/invalid_service_15/service.bal:43]:
```suggestion
        return "John";
```

Comment (DimuthuMadushan) [reply compiler-plugin-tests/src/test/resources/ballerina_sources/invalid_service_15/service.bal:43]: seems like every `service.bal` file has this extra space.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/validator/ServiceAnalysisTask.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (DimuthuMadushan) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/validator/ServiceAnalysisTask.java:2]:
```suggestion
 * Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/validator/ServiceValidator.java [hunk 4]
```diff
@@ -18,98 +18,387 @@
...
+        if (typeSymbol.getName().isEmpty()) {
+            return false;
+        }
+        if (!isGraphqlModuleSymbol(typeSymbol)) {
+            return false;
         }
-        return false;
+        return FILE_UPLOAD_IDENTIFIER.equals(typeSymbol.getName().get());
     }
 }
```
Comment (DimuthuMadushan) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/validator/ServiceValidator.java:404]: if we have a distinct service class and a record type as union-type members, do we allow them?

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/validator/ServiceAnalysisTask.java (hunk 3 above)
Comment (ThisaruGuruge) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/validator/ServiceAnalysisTask.java:2]: This is actually a file rename. GitHub incorrectly identified it as a file deletion + new file addition.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/validator/ServiceValidator.java (hunk 4 above)
Comment (ThisaruGuruge) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/graphql/compiler/validator/ServiceValidator.java:404]: No, we don't. Only distinct service classes are allowed.

File: compiler-plugin-tests/src/test/resources/ballerina_sources/invalid_service_15/service.bal (hunk 2 above)
Comment (ThisaruGuruge) [reply compiler-plugin-tests/src/test/resources/ballerina_sources/invalid_service_15/service.bal]: Fixed with e041ee09edb6ce8c61b72054d856473b78a6613e

**Review by DimuthuMadushan — APPROVED** [review]

**Pull request merged**